### ISSUE TITLE
Transition to Raspbian

### ISSUE THREAD

**Issue by atarijookie** [atarijookie/ce-atari#47]
Tin and tuxie are already looking into this...

Apart from creating the raspbian ce image, there is also need for:
- detecting the linux image type from code
- disable functions which rely on raspbian when running on yocto
- possibly start releasing 2 builds of ce main app - one for yocto, one for raspbian
- fix things which work differently on these two
- on update download the right update package
- add ce function, which could tell the ST apps if the linux image is raspbian or yocto, use that function in all st apps which need raspbian but got runned on yocto (mp3 playback app?)

**Comment by tin-nl** [comment]
The current plan is to keep a failsafe distribution (Yokto) on the stock SD card. Unfortunately the latest Jessie(s) switched the boot/kernel (aka the first partition with config.txt and the like) to an incompatible version. Yocto and ce_main itself runs, but e.g. cifs mount won't work anymore. Unfortunately there's not enough space on the boot partition of the Stock yocto image to keep two kernels there. I'm not even sure if the current kernel fits correctly, it seemed afwully tight.

So we have to decide if we should keep this fallback - for all that we know the next raspbian might introduce a kernel where yocto won't work at all. 
This also has the implication, that I think we have to follow @tuxie01 s idea to require an USB stick for upgrading the distribution (unfortunately the data on that stick will be destroyed). The whole SD card has to be re-written, including boot partition, because it's size will change (the kernel needs more space with alle the new PIs to support and all that).

If we take the USB stick approach, the only two things left solving is 
1) how to write a running distribution to the SD card - otherwise we have a bigger download with two dists (one running on USB, one as IMG for dd)
2) how to show progress to the user

Possible workflow:
1. user inserts USB key (formatted) with upgrade package on it
2. yoco detects package, creates a partition on SD card
3. yocto reboots (unfortunately yocto doesn't re read partition table in some cases)
4. yocto detects package and partition, copies package to partition
5. yocto dds package to USB stick
6. yocto saves CE config to USB key
6. yocto changes boot partition to USB key
7. reboot
8. USB key distribution dds itself to SD card
9. USB dist changes fstab etc. on SD card and copies config backup,to SD card
10. USB dist changes boot partition to SD card
11. reboot

That way we have only one component more if we'd like to have an internet update: dl update package to USB stick.

Thoughts?

**Comment by atarijookie** [comment]
OMFG, that's complicated, and that means it has more opportunities to get broken :-/ And as there is no real way to show progress of the whole, people will be inpatient, turn it off too soon screwing it somewhere in the middle. 

The easiest way to show the 'progress' would be just to connect RPi to TV using HDMI and watch as it reboots, does stuff and so on, but most of the people will be lazy to do that, so they would do it blindly, or won't do it at all. 

I would start with user saving his config manually - by opening web interface and downloading the settings, then we could kick out the saving and restoring of the config to / from USB key.

I would also force the user to download the img for dd on PC - that way he would see if it was downloaded completely, see the download progress and so on, even possibly let him unzip it so he would do a CRC check when ZIP is depacking.

The only issue I see and which I'm not sure if it can be done, and which you probably found out that it can't be done - writing the img through dd on the whole SD card, while the linux is running from that SD card. Is that (im)possible? 

The whole process seems too complicated to be worth implementing for only few people :(

**Comment by tin-nl** [comment]
> The only issue I see and which I'm not sure if it can be done, and which you probably found out that it can't be done - writing the img through dd on the whole SD card, while the linux is running from that SD card. Is that (im)possible?

With the cur yocto it's not possible, the system has to run from sth ot than the write target or it would crash. That might be ramfs or USB. The procedure stays almost the same.
That's a m current script uses your double buffer idea - create new dist on SD car and switch to that one. Unfortunately the PIS boot partition cna't b configured.

> The whole process seems too complicated to be worth implementing for only few people :(

That's the thing whe I woul be great to have some numbers... what's the ratio of WiFi image dls to devices? Ratio of updates to devices?
You migh be righ that this I just a few people - but is it?

We then need at least some strategy to disencourage fragmentation...

**Comment by tin-nl** [comment]
And we surely don't need to wait for this upgrade path to switch to raspbian. Except @tuxie01 s kernel (which might o might not be as fast booting as yoctos, last checked it was much slower, but those were first tries) and the dist detection, everything else should be a go so far (disregarding t build processes of course).

**Comment by atarijookie** [comment]
> That might be ramfs or USB

Then what about just adding some script, that would copy the minimal yocto to ramfs on the next boot, start that one instead of the one on the SD card, and then write the whole image from USB?

> That's the thing whe I woul be great to have some numbers

Well, I don't have any numbers... I was thinking about adding something for reporting the app version  and serial number to the server, so at least I could create a table like:
this serial number - that app version

That would tell us how many people are alive, and how many have the new app... But I'm worried that people would be pissed off if they would find out that I'm collecting some info about their device - you know, with all those cases of OS or some SW reporting stuff about you people are getting paranoid and also pissed... 

> Except tuxie01 s kernel...

Why do we need a custom kernel? I thought that one of the points of switching to standard raspbian was to have the standard kernel, so stuff would just work as it should...

**Comment by tin-nl** [comment]
> Then what about just adding some script, that would copy the minimal yocto to ramfs on the next boot, start that one instead of the one on the SD card, and then write the whole image from USB?

I guess Yoctos probably too big for that, but I'l check.

>  But I'm worried that people would be pissed off if they would find out that I'm collecting some info about their device

Yeah, I'll see. But I'll suspect that at least the Wifi image DLs should exist in some logfile somewhere? That would at least give us a ratio how many first-batch-people took the hard way to update.

> Why do we need a custom kernel? I thought that one of the points of switching to standard raspbian was to have the standard kernel, so stuff would just work as it should

Absolutely! We don't need a custom kernel - @tuxie01 just wants to see if he can generate a significantly faster time-to-network by cautiously removing kernel modules.

**Comment by miniupnp** [comment]
I'm sorry to ask, but what are the reasons for switching to raspbian, apart from being relatively easier to use for developers (which we shouldn't care as long as users are satisfied ;) ?

**Comment by atarijookie** [comment]
> what are the reasons for switching to raspbian
1. supporting RPi 2 and RPi 3 - users want to use the newer, more powerful devices, plus the original RPi B+ are now getting harder to buy as they are being replaced by RPi 2 and RPi 3
2. easier to add newer features based on libs and apps which are available on Raspbian, but aren't available on Yocto (mp3, video, vnc, fuse-zip, anything)
3. as mentioned - easier for developers - crosscompilation, newer tools, and so on

**Comment by mikrosk** [comment]
Wouldn't help here if we could boot straight from the usb? You know, the user would download & dd the image to his USB stick, put it into the CE, rebooted the device and let it do its magic.

**Comment by tin-nl** [comment]
Hello Mikro, nice to have you here! :-)
Booting to USB would be nice (and was considered of course), but unfortunately the kernel changed and the yocto kernel is too old to run certain drivers. The dist would boot, but unstable and is missing e.g. network and god knows what else is.

Also, "put it into the CE, rebooted the device and let it do its magic." won't work on PI (at least not this straight forward) - there is no USB autoboot, the bootloader has to be changed to the exact partition to be booted. That's of course possible but would "brick" the device for the average user if USB doesn't boot (see instabilities above).

Long story short: the kernel change does make the "multiple Distributions on on device" idea obsolete - at least with yocto as a fallback. Currently the best thing to do to dd Raspian to the SD card, unfortunately

Any ideas/experiences with this on your part?

**Comment by tin-nl** [comment]
Oops :-)

**Comment by mikrosk** [comment]
I still don't get it. One minute google results:

_Use dd or winimage to write the image to the USB stick rather than the sdcard.
Format the sdcard as FAT and copy the files from the FAT partition of the USB stick onto the sdcard (only needs to be about 64M in size)._

_Change cmdline.txt on sdcard from root=/dev/mmcblk0p2 to root=/dev/sda2_

Or even better: http://jonathanmh.com/boot-raspberry-pi-from-a-usb-stick (without formatting the SD card). Seems pretty kernel/distro agnostic to me, so what am I missing here? ;-)

**Comment by tin-nl** [comment]
Well, those google results don't handle custom dists like yocto, and they oversimplify (which is ok in most use cases).

The kernel is (always!) located on SD card on the boot partition and can't be loaded from USB stick. PI boots the GPU from that boot partition, which boots the CPU kernel which loads the device tree.
 The CE kernel is old and doesn't work well with current Raspian (which switched kernels in the last release or two). My main problem was mount AFAIR, and it's documented somewhere that this breaks with aforementioned kernel mismatch. So we would boot into a system where several things will be broken.

This means that the boot partition has to be rewritten with a new kernel (and the partition doesn't have enough space for a backup copy of the old kernel and modules) - which means that as soon as the USB stick is removed or sth else is wrong, yocto won't boot anymore due to the new kernel. The whole point of this exercise was to have a "core" kernel, that boots the CE if anything with the new kernels is 

Again: not unsolvable, but to get it completely right, we have to either write an update script that unfortunately can't go back if anythings going wrong or ask the user to remove and rewrite the card. The dist upgrade I did was recoverable to a working state by an average user w/o opening the CE or even knowing what a terminal is. Due to the new kernel and the resulting mismatch that won't work anymore.

But we could do a Yocto on an USB stick that rewrites the whole SD card with raspbian and a new boot partition.

I hope it's a bit clearer now? :-)

**Comment by mikrosk** [comment]
I see. I think the key point here is this: "we have to either write an update script that unfortunately can't go back if anythings going wrong or ask the user to remove and rewrite the card".

My school of thought ;-) says, yes, let's do it like that. If the user messes up the USB stick (pull it out, incorrectly copies it, ...) we can always say "dude, get a new usb stick, 'dd' it, plug into the CE and try again". And as you say, as the totally fall-back solution, the user can always pull the card out and overwrite it the the original system.

EDIT: Sorry I missed this: _"But we could do a Yocto on an USB stick that rewrites the whole SD card with raspbian and a new boot partition."_

Now I see there are two different scenarios:
1. Boot from the SD card, at some point switch to the USB stick, use its filesystem, do something
2. Boot from the SD card, initiate a script which replaces the whole SD card with new system (whatever that is).

What I'd like to achieve is something like "best of both worlds" -- i.e. making the copy process as dumb as possible but in the same time keep the old kernel alive as long as possible, in case the user pulls of the USB stick or a power failure or whatever. So by the next boot the OS on the SD card would detect that there's an update process in progress and expected an USB stick with the image to copy from. Do you think something like that is achievable?

**Comment by tin-nl** [comment]
> My school of thought ;-) says, yes, let's do it like that.

We would need a yocto (or at least a minimal dist with that kernel version used there) for usb, which handles the dd. And an option in the CE config to upgrade, which checks for the correct usb stick inserted and then changes the boot config to USB and reboots. 
@jookie: what's the easiest/low complexity way to send progress from CE to ST? A binary that just sends a ACSI sector to the ST and then exits should be feasible, shouldn't it?

**Comment by mikrosk** [comment]
After some thinking... what about this:
1. An update delivers a set of scripts.
2. A script will verify existence of a "recovery partition" and its content. If exists, skip to 5
3. Using scripted `fdisk` commands we create another partition.
4. Reboot && `mkfs.ext2` && download from Internet/USB the recovery image && `dd` to recovery partition
5. Set boot flag to the recovery partition
6. Reboot
7. Recovery startup script keeps looking for an image to download from Internet or an USB stick
8. `dd` image to the main partition
9. Verify the content of the main partition
10. Set boot flag to the main partition
11. Reboot

So the only weak point here is that the user switches off the devices during `fdisk` commands and that only once -- as soon there's a recovery partition, we're safe for any kind of update.

**Comment by tin-nl** [comment]
That's more or less what my upgrade script does, except there's no mkfs in CEs busybox/yocto (I use e.g. dd to "mkfs" an empty dl partition) and it dls the new dist, not a recovery. In the current state the workflow described above (May 4) works (san the kernel problem) but inplace, w/o need of USB stick. And the automatic restarting of the update script after reboot is missing. I have to revisit this regarding USB stick and the new kernel/replacing yocto with a new recovery dist. I can send you the cleaned up scripts in the next few days if they are of any interest.

**Comment by mikrosk** [comment]
_And the automatic restarting of the update script after reboot is missing_

But if the user switched off the ST during `dd`, the system is bricked, isn't it. But you're right, using `dd` instead of `mkfs` makes much more sense, we actually don't need to copy anything, just verify the result.

Sure, I'd love to see the scripts, go ahead!

**Comment by atarijookie** [comment]
> what's the easiest/low complexity way to send progress from CE to ST? A binary that just sends a ACSI sector to the ST and then exits should be feasible, shouldn't it?

Depends if you want to send one progress update, or send more of them in a cycle, but that what you describe should be it. Could be based on the simple-boot / fast-boot binary...

**Comment by atarijookie** [comment]
Btw., this update thing from one to another is either slowing the whole transition down, or we (I) don't need to wait for it :) Before the next batch of devices I need to get Raspbian working with it, even if it would be only remove_SD - write_image - reinsert_SD style, because the older RPi's are now hard to get, and the Yocto doesn't work with the new ones.

**Comment by mikrosk** [comment]
Yeah, those are two separate things (and I admit it may be my fault of hijacking this thread for the update process). Making Raspbian working on the CE has definitely higher priority than a nice update process.

**Comment by miniupnp** [comment]
isn't it possible to buiid a new kernel for the yocto ?

**Comment by mikrosk** [comment]
I guess you've seen this already: https://www.raspberrypi.org/blog/pi-3-booting-part-i-usb-mass-storage-boot ... wont help the old units but maybe new devices could be equipped with the RPi3.

**Comment by atarijookie** [comment]
OK, I'm back from dead (probably). No activity here for 4 months, so for now screw the nice update from Yocto to Rapsbian, I need to make it work on Raspbian without the nice update option now. 

@tin-nl I remember you mentioned that you already are using CE under Raspbian - do you need to use newer version of the bcm2835 library? (I guess so, as the newer seems to support RPi 2, and older does not). Does the libcivetweb needs to be rebuilt for Raspbian, or one build can be shared between Yocto and Raspbian?

**Comment by tin-nl** [comment]
@atarijookie: I'm currently in summer hiatus, too - let me check the details in the next few days. From the top of my head: I think I didn't try to run the raspbian build with yocto, only t he other way around.

**Comment by tin-nl** [comment]
@atarijookie: I used the repositories libbcm2835-1.35. Did not Test on Pi2 though.
The libcivitweb.a that's checked in is used as is.

**Comment by tin-nl** [comment]
@atarijookie: regarding quicker startup under raspian - the easiest way to achieve this is to start comsosex main app via init.d and add it to the appropriate runlevels.

Here's a gist with the needed files to do that:
/ce/init.d/cosmosex (Start/Stop script)
/ce/ce_boot.sh (Startup script)
/ce/cesuper.sh (slightly modified watcher)

https://gist.github.com/tin-nl/62ff6131fc30b8dc2e0ff5959b576932

Link /ce/init.d/cosmosex into /etc/rcS.d/S09cosmosex and /etc/rc6.d/K01cosmosex and the main app should boot at a reasonable speeed (for standard raspbian).

**Comment by mikrosk** [comment]
Hmm, have you heard of this? http://www.pibakery.org ... maybe it could be of a help?

**Comment by atarijookie** [comment]
@tin-nl Thank, I'll take a look ad that.

@mikrosk Yes, saw it, had a run, but in its current state it lacks some good templates / steps which you could use - it's mostly empty, and your options to add there through the UI are like 'run this command' or 'install this app' - and we don't need a UI app to do that :)

**Comment by atarijookie** [comment]
@tin-nl -- in the gist mentioned above, I'm taking a look at the cesuper.sh, and the only difference I see is that it uses '-e' to check for the existence of the update script instead of '-f' -- is that all, or am I missing something? And is it better to use '-e' instead of '-f'? (just asking, I have no idea :) )

**Comment by miniupnp** [comment]
-e will work even if the file is not a standard file (a symbolic link for example)

**Comment by atarijookie** [comment]
@miniupnp thanks :) I thought that using -f for checking if it's a regular file would be better, but I didn't consider the possible existence of symlinks, so for that case it might be better :+1:

**Comment by tin-nl** [comment]
Well, I'm late to the party, it seems :) Of course @miniupnp is absolutely right - I can't remember the exact reason I did that. I probably changed sth in the directory structure and used links in there.

**Comment by atarijookie** [comment]
Transition without direct update from yocto done, thanks to everyone who helped. :+1:
